File: src/semantic/DimensionalChecker.cpp
#pragma once

#include "olang/frontend/ASTType.hpp"
#include "olang/semantic/DimensionalChecker.hpp"
#include <unordered_map>
#include <string>
#include <memory>
#include <vector>
#include <optional>

namespace olang::semantic {

  // ───────────────────────────────────────────────────────────────────────────
// Dimensional Analysis Checker (Biological Safety)
// ───────────────────────────────────────────────────────────────────────────
class DimensionalChecker {
public:
    explicit DimensionalChecker(ErrorReporter& reporter) 
        : errorReporter_(reporter) {}
    
    // Check if operation is dimensionally valid
    bool checkBinaryOp(const frontend::Type& lhs, const frontend::Type& rhs, 
                      const std::string& op);
    
    // Check if units match for comparison/addition/subtraction
    bool checkSameDimensions(const frontend::Type& t1, const frontend::Type& t2);
    
    // Compute result type for multiplication/division
    std::unique_ptr<frontend::Type> computeResultType(
        const frontend::Type& lhs, const frontend::Type& rhs, const std::string& op);
    
private:
    ErrorReporter& errorReporter_;
    
    void reportDimensionalError(const std::string& message, 
                               size_t line = 0, size_t col = 0);
};

} // namespace olang::semantic

File: src/semantic/dimensional_checker.rs
use crate::parser::ast::Type;
use super::error_reporter::ErrorReporter;

pub struct DimensionalChecker<'a> {
    reporter: &'a mut ErrorReporter,
}

impl<'a> DimensionalChecker<'a> {
    pub fn new(reporter: &'a mut ErrorReporter) -> Self { Self { reporter } }

    /// Check whether a binary operation is dimensionally valid.
    pub fn check_binary_op(&mut self, lhs: &Type, rhs: &Type, op: &str) -> bool {
        if op == "+" || op == "-" {
            self.check_same_dimensions(lhs, rhs)
        } else {
            // Multiplication and division combine dimensions — always permitted.
            true
        }
    }

    /// Addition/subtraction requires identical types (simplified).
    pub fn check_same_dimensions(&mut self, t1: &Type, t2: &Type) -> bool {
        if t1 == t2 { return true; }
        self.reporter.error(format!(
            "dimensional mismatch: '{}' and '{}' have incompatible dimensions",
            t1.to_display(), t2.to_display()
        ));
        false
    }

    /// Compute the result type for a binary arithmetic operation (simplified).
    pub fn compute_result_type(&self, lhs: &Type, _rhs: &Type, _op: &str) -> Type {
        lhs.clone()
    }
}
